src/JSInterop/Microsoft.JSInterop/src/IJSRuntime.cs

```diff
@@ -22,7 +22,7 @@ public interface IJSRuntime
         /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>window.someScope.someFunction</c>.</param>
         /// <param name="args">JSON-serializable arguments.</param>
         /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
-        ValueTask<TValue> InvokeAsync<TValue>(string identifier, object?[]? args);
+        ValueTask<TValue?> InvokeAsync<TValue>(string identifier, object?[]? args);
```

Comment:
I'm not sure about this. Why would we want to make it a rule that developers have to allow for `null` return values? Why can't developers make their own choice about whether their JS-side code might possibly return `null` or not?

It makes a bunch of our own code more messy, adding null checks for things that we know for sure are never null. Do we definitely want to impose this on everyone else too?

Or is there some reason I'm missing about why we have no choice but to do this?

Reply:
My guess is that your reasoning is that the value *might* in fact be `null`, because JS might return null/undefined, and then STJ will give you `null` at runtime. However that's no more true when calling JS code than it is when calling .NET code (which might have not-null annotation but then actually returns null at runtime).

The best .NET-only analogy for JS interop would be a developer making calls to their own code (because both the .NET and JS parts are typically the developer's own code). When calling your own code, you get to choose what the nullability annotation says on the called-code side, and it's up to you the developer to get that right. If you're wrong, and you say something returns nonnull but it actually returns null at runtime, then you'll get a NRE when dereferencing but that's fine because you shouldn't get it wrong. In this case because the called-code is JS you can't put the nullability annotation there, so you have to put it in the generic parameter to `InvokeAsync` instead. Why shouldn't you be free to specify whatever nullability you want there? If the value might be null, say so, and if not, say not.

The only case I could make for treating .NET specially is that JS return values are more untrusted than .NET return values. If there are reasonable security reasons for needing to check for null return values then I would agree it's helpful to make the compiler guide people to do that. However I can't really think of cases where it would be *more* problematic to get a NRE during dereferencing compared with throwing another exception preemptively due to a null check.

Reply:
Another argument in favour of allowing `TValue` and not forcing `TValue?` is that, if we really insist on saying things might be null when they come back from JS, then we'd have to do the same for all the properties on the instance, not just the instance itself.

That is, STJ would have to have the rule that your `TValue` contains only nullable properties, and the same thing recursively for those property types. But we don't do that for the properties, so why do it for the top-level instance?

Reply:
I agree, making this nullable does make it a hassle to use. We have in the past made quality of life choices with nulllability in our APIs, we could leave these be. I'll send an update.

Reply:
Thanks! I've rescinded the "request updates" review flag to avoid blocking this.